File: DSA_PYTHON/Recursion/TaylorSeries.py
class Exponential:
    def __init__(self):
        self.p = 1
        self.f = 1
        self.s = 1

    def e(self, x, n):
        if n == 0:
            return 1
        r = self.e(x, n - 1)
        self.p *= x
        self.f *= n
        return r + self.p / self.f

    # second version of doing "horner's rule"
    def ev2(self, x, n):
        if n == 0:
            return self.s
        self.s = 1 + x * self.s / n
        return self.ev2(x, n - 1)

    def main(self):
        print("------------------------------------")
        x = int(input("Enter 'x': "))
        n = int(input("Enter 'N': "))

        res = self.ev2(x, n)

        print(f"e^x is: {res}")
        return str(res)

File: DSA_CSHARP/Recursion/TaylorSeries.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DSA_CSHARP.Recursion;
internal class TaylorSeries
{
    double p = 1, f = 1, s = 1;
    double e(int x, int n)
    {
        double r;
        if (n == 0) return 1;
        r = e(x, n - 1);
        p = p * x; f = f * n;
        return r + p / f;
    }
    // second version of doing "horner's rule"
    double ev2(int x, int n)
    {
        if (n == 0) return s;
        s = 1 + x * s / n;
        return ev2(x, n - 1);
    }
    //without using recursion
    double eNoRecursion(int x, int n)
    {
        double s = 1, num = 1, den = 1;
        for (int i = 0; i <= n; i++)
        {
            num *= x;
            den *= i;
            s += num / den;
        }
        return s;
    }
    public string Main()
    {
        Console.WriteLine($"------------------------------------");
        Console.Write($"Enter 'x': ");
        var x = Convert.ToInt32(Console.ReadLine());

        Console.Write($"Enter 'N': ");
        var n = Convert.ToInt32(Console.ReadLine());

        var res = ev2(x, n);

        Console.WriteLine($"e^x is: {res}");
        return res.ToString();

    }
}
